<launch>

    <!-- what config we are going to run (should match folder name) -->
    <arg name="verbosity"   default="INFO" /> <!-- ALL, DEBUG, INFO, WARNING, ERROR, SILENT -->
    <arg name="config"      default="rs435i" /> <!-- euroc_mav, tum_vi, rpng_aruco -->
    <arg name="config_path" default="$(find ov_msckf)/../config/$(arg config)/estimator_config.yaml" />

    <!-- mono or stereo and what ros bag to play -->
    <arg name="max_cameras" default="2" />
    <arg name="use_stereo"  default="true" />
    <arg name="bag_start"   default="0" /> <!-- v1-2: 0, mh1: 40, mh2: 35, mh3: 17.5, mh4-5: 15 -->
    <arg name="bag_rate"    default="1" />
    <arg name="dataset"     default="V1_01_easy" /> <!-- V1_01_easy, V1_02_medium, V2_02_medium -->
    <arg name="dobag"       default="false" /> <!-- if we should play back the bag -->
    <arg name="bag"         default="/media/patrick/RPNG\ FLASH\ 3/$(arg config)/$(arg dataset).bag" />
<!--    <arg name="bag"         default="/home/patrick/datasets/$(arg config)/$(arg dataset).bag" />-->
<!--    <arg name="bag"         default="/datasets/$(arg config)/$(arg dataset).bag" />-->

    <!-- saving trajectory path and timing information -->
    <arg name="dosave"      default="false" />
    <arg name="dotime"      default="false" />
    <arg name="path_est"    default="/tmp/traj_estimate.txt" />
    <arg name="path_time"   default="/tmp/traj_timing.txt" />

    <!-- if we should viz the groundtruth -->
    <arg name="dolivetraj"  default="false" />
    <arg name="path_gt"     default="$(find ov_data)/$(arg config)/$(arg dataset).txt" />


    <!-- MASTER NODE! -->
<!--    <node name="ov_msckf" pkg="ov_msckf" type="run_subscribe_msckf" output="screen" clear_params="true" required="true" launch-prefix="gdb -ex run &#45;&#45;args">-->
    <!--ode name="ov_msckf" pkg="ov_msckf" type="run_subscribe_msckf" output="screen" clear_params="true" required="true">

        <param name="verbosity"              type="string" value="$(arg verbosity)" />
        <param name="config_path"            type="string" value="$(arg config_path)" />
        
        <param name="use_stereo"             type="bool"   value="$(arg use_stereo)" />
        <param name="max_cameras"            type="int"    value="$(arg max_cameras)" />
        
        <param name="record_timing_information"   type="bool"   value="$(arg dotime)" />
        <param name="record_timing_filepath"      type="string" value="$(arg path_time)" />

    </node-->

    <!-- play the dataset -->
    <group if="$(arg dobag)">
        <node pkg="rosbag" type="play" name="rosbag" args="-d 1 -r $(arg bag_rate) -s $(arg bag_start) $(arg bag)" required="true"/>
    </group>

    <!-- record the trajectory if enabled -->
    <group if="$(arg dosave)">
        <node name="recorder_estimate" pkg="ov_eval" type="pose_to_file" output="screen" required="true">
            <param name="topic"      type="str" value="/ov_msckf/poseimu" />
            <param name="topic_type" type="str" value="PoseWithCovarianceStamped" />
            <param name="output"     type="str" value="$(arg path_est)" />
        </node>
    </group>

    <!-- path viz of aligned gt -->
    <group if="$(arg dolivetraj)">
        <node name="live_align_trajectory" pkg="ov_eval" type="live_align_trajectory" output="log" clear_params="true">
            <param name="alignment_type" type="str" value="posyaw" />
            <param name="path_gt"        type="str" value="$(arg path_gt)" />
        </node>
    </group>

    <node pkg="nodelet" type="nodelet" args="manager"
    name="ovmsckf_nodelet_manager" output="screen">
        
    </node>

	
	<node pkg="nodelet" type="nodelet" args="load ovmsckf_nodelet/OvmsckfNodeletClass ovmsckf_nodelet_manager"
    name="OvmsckfNodeletClass_loader" output="screen">
        <param name="config_path"            type="string" value="$(arg config_path)" />
        <param name="publish_global_to_imu_tf" type="bool" value="false"/>
        <param name="publish_world_to_body_tf" type="bool" value="false"/>
        <param name="publish_calibration_tf" type="bool" value="false"/>
        <!-- <remap from="/zedm/zed_node/imu/data_raw" to="/zed_nodelet/imu/data_raw" />
        <remap from="/zedm/zed_node/left_raw/image_raw_gray" to="/zed_nodelet/left_raw/image_raw_gray" />
        <remap from="/zedm/zed_node/right_raw/image_raw_gray" to="/zed_nodelet/right_raw/image_raw_gray" /> -->
        <!-- remap from="/OvmsckfNodeletClass_loader/odomworld" to="/OvmsckfNodeletClass_loader/odomworld_laptop" / -->
	</node>

    <arg name="tf_namespace"              default="drone31"/>
    <arg name="device_type_camera_f"      default="d4.5"/>
    <arg name="camera_f"                  default="cam1"/>              <!-- Note: Replace with camera name -->
    <arg name="clip_distance"             default="-2"/>
    <arg name="tf_prefix_camera_f"        default="$(arg tf_namespace)/$(arg camera_f)"/>

    <arg name="initial_reset"             default="true"/>
    <arg name="enable_fisheye"            default="false"/>
    <arg name="color_width"               default="640"/>
    <arg name="color_height"              default="480"/>
    <arg name="depth_width"               default="640"/>
    <arg name="depth_height"              default="480"/>
    <arg name="topic_odom_in"             default="odom_in"/>
    <arg name="calib_odom_file"           default=""/>
    <arg name="d435_imu_unite"            default="linear_interpolation"/>
    <arg name="t265_imu_unite"            default="copy"/>
    <arg name="gyro_fps_d435"             default="400"/>
    <arg name="accel_fps_d435"            default="250"/>
    
    <param name="/cam1/stereo_module/exposure"        value="2500"/>
    <param name="/cam1/stereo_module/emitter_enabled" value="0"/>

    <group ns="$(arg camera_f)">
      <include file="$(find realsense2_camera)/launch/includes/nodelet.launch.xml">
        <arg name="device_type"                 value="$(arg device_type_camera_f)"/>
        <arg name="tf_prefix"                   value="$(arg tf_prefix_camera_f)"/>
        <arg name="initial_reset"               value="$(arg initial_reset)"/>
        <arg name="align_depth"                 value="false"/>
        <arg name="enable_infra1"               value="true"/>
        <arg name="enable_infra2"               value="true"/>
        <arg name="enable_accel"                value="true"/>
        <arg name="enable_gyro"                 value="true"/>
        <arg name="enable_depth"                value="false"/>
        <arg name="color_width"                 value="$(arg color_width)"/>
        <arg name="color_height"                value="$(arg color_height)"/>
        <arg name="depth_width"                 value="$(arg depth_width)"/>
        <arg name="depth_height"                value="$(arg depth_height)"/>
        <arg name="clip_distance"               value="$(arg clip_distance)"/>
        <arg name="unite_imu_method"            value="$(arg d435_imu_unite)"/>
        <arg name="accel_fps"                   value="$(arg accel_fps_d435)"/>
        <arg name="gyro_fps"                    value="$(arg gyro_fps_d435)"/>
        <arg name="stereo_module/exposure/1"    value="2500"/>
        <arg name="stereo_module/gain/1"        value="16"/>
        <!-- <arg name="external_manager"            value="true"/> -->
        <!-- <arg name="manager"                     value="ovmsckf_nodelet_manager"/> -->
      </include>
    </group>
	
    <node pkg="nodelet" type="nodelet" args="load odom_transform_nodelet/OvtransformNodeletClass ovmsckf_nodelet_manager"
    name="odom_transform_nodelet" output="screen">
     <param name="config_path"            type="string" value="$(find odom_transform)/config/transform.yaml" />
        <rosparam file="$(find odom_transform)/config/transform.yaml" />
	</node>

</launch>
